### PR TITLE
fix(ffi)!: replace blocking bootstrap provider FFI call with push-based receivers

### LINKED ISSUE #686: Bootstrap provider FFI: unbounded buffering and a blocking call on async workers
The bootstrap provider crossing is the last FFI path built on a synchronous run-to-completion call: the host invokes the plugin's `bootstrap_fn` inline from an async task, and it only returns once the whole snapshot has been handed over. Two problems follow.

1. Unbounded memory. The adapters on both sides of the call use unbounded std channels (the host bridge in `host-sdk/src/proxies/bootstrap_provider.rs`, the plugin shim in `plugin-sdk/src/ffi/vtable_gen.rs`). The query-side channel is bounded at 1000, so a slow consumer pushes the entire snapshot into those links. Measured: 50k events fully buffered in-process with the consumer stalled, about 67 MB of RSS. Bounding either link alone does not help, the backlog just moves to the other one.

2. Worker starvation. The inline call pins an async worker for the whole bootstrap. Two concurrent bootstraps on a 2-worker runtime stalled unrelated tasks for 2 seconds. Bounding the bridge without also moving the call off the workers deadlocks the runtime once concurrent bootstraps reach the worker count.

Rather than patching this shape (bounded channels plus blocking pool), replace the contract with the push pattern the codebase already uses for CDC and for source-plugin bootstrap: `FfiBootstrapReceiver` for events, `FfiBootstrapResultReceiver` for completion. `bootstrap_fn` returns immediately, events flow through a bounded sync channel whose callback blocks when full, and the result arrives through the existing callback with its error semantics. This removes the blocking call, the per-bootstrap forwarding threads, and every unbounded link, and cancellation becomes dropping the receiver.

Memory expectation, to be precise: this bounds what the transport can buffer. In-flight events per bootstrap stream become a small constant set by the channel capacities, regardless of snapshot size or consumer speed, so a stalled query no longer pulls the whole snapshot into memory. Peak usage can still spike with data size inside providers that materialize their snapshot before sending (the postgres bootstrapper buffers each full table, oracle the whole snapshot). That is provider work, tracked in #690.

Out of scope: BootstrapContext properties are legacy and unused - oracle source has a bug in using them. These do not cross the FFI boundary even without the change, so they will remain unused. Cleaning them up is not in scope of this issue.

### PR DESCRIPTION
**Pull request by amansinghoriginal**
Closes #686.

Implements the receiver contract from the issue: bootstrap_fn returns FfiBootstrapReceiver plus FfiBootstrapResultReceiver immediately, every link on the path is bounded, and FfiBootstrapSender is gone. One shared consumer module serves both the host proxy and the plugin-side proxy. The provider wrapper is now Arc-shared so dropping the vtable during an in-flight bootstrap cannot free it early. Error paths release partially-initialized streams instead of leaking them, producer drop functions run behind panic barriers, failed thread spawns deliver the stream-end sentinel or result inline, and the host proxy starts its consumers on a dedicated thread, matching the subscribe path.

New integration tests drive the real producer and both consumers in-process: a stalled consumer bounds in-flight events and stalls the provider, concurrent bootstraps complete on a 2-worker runtime, provider errors propagate, source_position round-trips, cancellation stops the provider, and a double-hop chain stays bounded and lossless. The scriptfile bootstrap plugin also builds as a test cdylib now, so streaming and backpressure are verified across a real dylib boundary.

ABI break: BootstrapProviderVtable changed and FFI_SDK_VERSION is now 0.13.0 (0.12.0 was taken by the log-level ABI change in #689), so hosts reject plugins built against the old layout at load time. All existing 0.11.0 and 0.12.0 plugin binaries need a rebuild.

### CODE REVIEW

File: components/plugin-sdk/src/ffi/bootstrap_proxy.rs [hunk 1]
```diff
@@ -35,154 +39,85 @@ pub struct FfiBootstrapProviderProxy {
...
-                None
-            } else {
-                Some(unsafe { *Box::from_raw(ptr) })
+            let stream = unsafe { *Box::from_raw(stream_ptr) };
+            if stream.events.is_null() || stream.result.is_null() {
+                anyhow::bail!("Bootstrap provider returned an incomplete stream");
             }
-        });
+            let events = unsafe { *Box::from_raw(stream.events) };
+            let result = unsafe { *Box::from_raw(stream.result) };
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_proxy.rs:120]: Fixed in the latest push. Both consumers now release whichever receiver is populated via `release_bootstrap_receiver`/`release_result_receiver` before bailing. This also closes the provider's event channel, so the provider thread parked behind it exits instead of leaking along with the memory.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+unsafe impl Send for FfiReceiverState {}
+unsafe impl Sync for FfiReceiverState {}
+
+impl Drop for FfiReceiverState {
+    fn drop(&mut self) {
+        (self.drop_fn)(self.state);
+    }
+}
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs:168]: Added in the latest push: `guarded_producer_drop` wraps the call in `catch_unwind`. Worth noting the exposure was already narrow: our own producers guard their drop fns internally with `ffi_guard`, and since Rust 1.81 an unwind escaping an `extern "C"` definition aborts on the callee side, so this barrier is defense in depth for foreign producers rather than a fix for reachable UB.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    }));
+    (inner.start_fn)(
+        inner.state,
+        bootstrap_result_callback,
+        ctx as *mut std::ffi::c_void,
+    );
+    // The producer's start_fn moves the pending result out of its state, so
+    // the state can be released immediately (matches the subscribe path).
+    (inner.drop_fn)(inner.state);
+    rx
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs]: Same fix as the drop-site above: the call now goes through `guarded_producer_drop`.

File: components/host-sdk/src/proxies/bootstrap_provider.rs [hunk 4]
```diff
@@ -61,93 +61,70 @@ impl BootstrapProvider for BootstrapProviderProxy {
...
+                ));
+            }
+            let stream = unsafe { *Box::from_raw(stream_ptr) };
+            if stream.events.is_null() || stream.result.is_null() {
+                return Err(anyhow::anyhow!(
+                    "Bootstrap provider returned an incomplete stream"
+                ));
+            }
+            let events = unsafe { *Box::from_raw(stream.events) };
+            let result = unsafe { *Box::from_raw(stream.result) };
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/host-sdk/src/proxies/bootstrap_provider.rs:112]: Fixed together with the plugin-side path: both consumers share the same release helpers from `bootstrap_stream`, and the incomplete-stream handling in each proxy covers this error path.

**Comment by amansinghoriginal** [comment]
> ### 🟡 Not ready to approve
> The new bootstrap stream consumer contains a `#[must_use]` value that is currently ignored and can produce an `unused_must_use` warning (potentially breaking CI if warnings are denied).
> 
> _Once you've addressed the issues Copilot identified, you can request another Copilot review._

Re: the latest Copilot review's suppressed `#[must_use]` comment on `bootstrap_stream.rs` --> that's a false positive. Neither `Arc` nor `Arc::from_raw` carries `#[must_use]`; the exact pattern compiles clean even under `#![deny(unused_must_use)]`, and clippy `--all-targets` on this branch is warning-free. Letting the reconstructed `Arc` drop at end of statement is the intended single-point reclaim of the leaked callback reference (see the comment above it), and matches the existing idiom in `host-sdk/src/proxies/change_receiver.rs`.

**Review by danielgerlag — CHANGES_REQUESTED** [review]
## Review: PR #691 — push-based bootstrap provider FFI

Overall design looks right for #686: immediate `bootstrap_fn` return, bounded end-to-end path, shared consumer module, Arc-shared provider, ABI bump to `0.12.0`, and solid in-process + cdylib coverage. Not merge-ready until the spawn-failure double-callback bugs are fixed.

### Blockers

1. **`stream_start_push` double callback / UAF on spawn failure** (`components/plugin-sdk/src/ffi/vtable_gen.rs`)
   - When `thread::Builder::spawn` returns `Err`, Rust **drops the closure**. That runs `SentinelOnDrop`, which already fires `callback(ctx, null)`.
   - The `Err` arm then fires the same null sentinel again.
   - Consumer path does `Arc::from_raw` on the null sentinel → second call is a double-free of the push callback context.
   - Verified locally that spawn `Err` drops the closure (Drop counter increments).
   - **Fix:** exactly-once delivery on spawn failure — rely on closure drop **or** fire inline, never both.

2. **`result_start` same double callback / UAF** (`components/plugin-sdk/src/ffi/vtable_gen.rs`)
   - Same pattern with `CallOnce`: spawn failure drops the closure → `CallOnce::drop` fires null result → inline arm fires again.
   - `bootstrap_result_callback` does `Box::from_raw(ctx)` on first entry → second call is UAF/double-free of `ResultCallbackContext`.
   - Same fix as above.

### Should-fix

3. **Null sentinel is not idempotent** (`bootstrap_stream.rs`)
   - `Arc::from_raw` on null event assumes exactly-once sentinel. Broken today by (1) and by a buggy producer.
   - Make reclaim once-only (atomic/once flag or take under mutex). Apply same discipline to result callback `Box::from_raw(ctx)`.

4. **Missing null `ctx` guard on event push callback** (`bootstrap_stream.rs`)
   - `bootstrap_result_callback` null-checks `ctx`; event path dereferences immediately. Add the same guard.

5. **Missing tests for cleanup / null-result paths**
   - `release_bootstrap_receiver` / `release_result_receiver` (partial stream cleanup) uncovered.
   - Null-result path ("Bootstrap provider ended without a result") uncovered (current error test only hits negative `event_count`).

### Nits (non-blocking)

- Export / share `PROVIDER_EVENT_CAPACITY` with tests instead of hardcoding `100`.
- Plateau detection via 2s stability can flake under heavy CI load.
- Host proxy `thread::spawn(...).join()` inside async still briefly pins a worker; OK if intentional for macOS TLS (document), else prefer `spawn_blocking`.
- Minor wording in `lib/src/bootstrap/agents.md` ("cancellation is dropping the receiver").

### What’s good

- Correct ABI break + `FfiBootstrapSender` removal.
- Shared `bootstrap_stream` for host and plugin-side proxies.
- Strong tests: stalled backpressure bound, concurrent bootstraps on 2 workers, cancel, error text round-trip, double-hop, scriptfile cdylib path.
- Partial-stream leak fix and producer-drop panic barriers look good.

**Verdict:** request changes for blockers (1)(2); should-fix (3)–(5) before merge preferred.

File: components/plugin-sdk/src/ffi/vtable_gen.rs [hunk 5]
```diff
@@ -2094,153 +2119,326 @@ pub fn build_reaction_vtable_from_boxed(
...
-                let _ = result_tx.send(bootstrap_result.map_err(|e| format!("{e:#}")));
-            })
+            if let Err(e) = spawned {
+                // The closure never ran (and was dropped, closing the event
+                // receiver), so its sentinel will never fire. Deliver it inline
+                // so the consumer sees a clean end-of-stream instead of waiting
+                // forever on a forwarder that cannot start.
+                log::error!("Failed to spawn bootstrap forwarder thread: {e}");
+                let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                    callback(ctx_raw as *mut c_void, std::ptr::null_mut());
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/vtable_gen.rs:2225]: I don't think this double fires. SentinelOnDrop is a local created inside the closure body, not a capture. When spawn fails the closure is dropped without ever running, so only its captures are dropped (ctx_raw, callback, and rx, which closes the channel). The sentinel struct never exists at that point, so the inline call is the only delivery. Easy to check: put a drop-instrumented struct inside a closure body and drop the closure without calling it, its drop never runs. A counter on a captured value does increment, which I suspect is what the local verification measured.

Removing the inline call would actually reintroduce the hang this arm exists to prevent: spawn fails, nothing fires, and the consumer waits forever since there is no timeout on the bootstrap path.

That said, the underlying worry is now moot either way. The sentinel reclaim in bootstrap_stream.rs is once-guarded in the latest revision, so even a hypothetical duplicate sentinel is a harmless no-op. I also reworded the comment here, since two reviewers read it the same way it clearly wasn't doing its job.

File: components/plugin-sdk/src/ffi/vtable_gen.rs [hunk 6]
```diff
@@ -2094,153 +2119,326 @@ pub fn build_reaction_vtable_from_boxed(
...
+                        };
+                        call.fire(result_ptr);
+                    });
+            if let Err(e) = spawned {
+                // The closure never ran (and was dropped), so its CallOnce
+                // guard will never fire. Deliver the null result inline so the
+                // consumer resolves to an error instead of waiting forever.
+                log::error!("Failed to spawn bootstrap result thread: {e}");
+                let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
+                    callback(ctx_raw as *mut c_void, std::ptr::null_mut());
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/vtable_gen.rs:2350]: Same situation as the forwarder thread above: CallOnce is a body local, not a capture, so a failed spawn drops only the captures and CallOnce never exists. The inline null delivery is the single firing, and removing it would leave the oneshot pending forever.

The result context is also reclaim-once now, same discipline as the event side, with a consumer-held guard keeping it alive until the result is read. So even if a producer did deliver twice, it's a no-op rather than a double free.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+        bootstrap_push_callback_inner(ctx, event)
+    }))
+    .unwrap_or(false)
+}
+
+fn bootstrap_push_callback_inner(
+    ctx: *mut std::ffi::c_void,
+    event: *mut FfiBootstrapEvent,
+) -> bool {
+    let context = unsafe { &*(ctx as *const PushCallbackContext) };
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs:95]: Done, added the null guard in the latest push.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+        // Forwarder-exit sentinel (guaranteed exactly once on forwarder exit).
+        // This is the SOLE point at which the leaked Arc reference is
+        // reclaimed, so the context outlives every event callback; reclaiming
+        // on any other return path too would double-free it. Tolerate a
+        // poisoned lock so the reclaim below always runs.
+        if let Ok(mut guard) = context.tx.lock() {
+            *guard = None;
+        }
+        context.notify.notify_one();
+        unsafe { Arc::from_raw(ctx as *const PushCallbackContext) };
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs]: Done. The sentinel reclaim is once-guarded now, and the result context got the same treatment with a consumer-held guard keeping it alive until the result is read.

File: components/plugin-sdk/src/ffi/bootstrap_proxy.rs [hunk 9]
```diff
@@ -35,154 +42,107 @@ pub struct FfiBootstrapProviderProxy {
...
-        let source_position = if let Some(ref r) = ffi_result {
-            if !r.source_position_ptr.is_null() && r.source_position_len > 0 {
-                let bytes = unsafe {
-                    std::slice::from_raw_parts(r.source_position_ptr, r.source_position_len)
-                };
-                let owned = bytes::Bytes::copy_from_slice(bytes);
-                // Free the source position allocation
-                if let Some(drop_fn) = r.source_position_drop_fn {
-                    (drop_fn)(r.source_position_ptr as *mut u8, r.source_position_len);
+            let stream = unsafe { *Box::from_raw(stream_ptr) };
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_proxy.rs:107]: Both proxies already consume through one shared implementation, bootstrap_stream.rs. BootstrapProviderProxy in host-sdk and FfiBootstrapProviderProxy here both go through BootstrapStreamConsumer and wrap_result_receiver, so the null ctx guard and the once-guarded reclaim land in both paths by construction. There's no second copy to keep in sync.

File: components/host-sdk/tests/bootstrap_provider_ffi_test.rs [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+    let (tx, rx) = tokio::sync::mpsc::channel(EVENT_CHANNEL_CAP);
+    let consumed = Arc::new(AtomicUsize::new(0));
+    let (gate_tx, gate_rx) = tokio::sync::watch::channel(false);
+    let consumer = tokio::spawn(gated_consumer(rx, consumed.clone(), gate_rx));
+
+    let boot = tokio::spawn(async move { proxy.bootstrap(req(), &ctx(), tx, None).await });
+
+    // Wait for production to plateau (stable over 2 seconds).
+    let mut last = 0usize;
+    let mut stable_since = Instant::now();
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/host-sdk/tests/bootstrap_provider_ffi_test.rs:194]: Fair concern about the wall clock dependency, but the suggested variant would hang the test in exactly the passing case: with backpressure working, produced plateaus below MAX_IN_FLIGHT because the bound includes slack (we see about 1.35k against a bound of 1.37k), so a wait for produced >= MAX_IN_FLIGHT never completes. The current loop also resets its 2s stability window whenever the count changes, so a slow runner just waits longer instead of asserting early. The 60s cap is the only hard limit. I'd keep it as is unless it actually flakes in practice.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+        let mut count = 0usize;
+        loop {
+            match self.rx.try_recv() {
+                Ok(event) => {
+                    if tx.send(event).await.is_err() {
+                        return count;
+                    }
+                    count += 1;
+                }
+                Err(std::sync::mpsc::TryRecvError::Empty) => {
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs:249]: I'd rather not take a new dependency for this one. The sync_channel plus Notify shape is the same pattern the CDC ChangeReceiverProxy and the source-plugin BootstrapReceiverProxy already use, so switching just this module would leave the codebase with two conventions for the same bridge. And since plugin-sdk is statically linked into every plugin, each new dependency becomes audit surface for the whole plugin ecosystem, which feels like a lot to pay for saving 15 lines. If flume is worth adopting, I think it should be a standalone refactor across all three consumers. Happy to file a follow-up issue for that if you'd like.

File: lib/src/bootstrap/agents.md [hunk 12]
```diff
@@ -15,16 +15,24 @@ mediated by the host.
...
   → Plugin B stores vtable, calls vtable.bootstrap_fn() when subscribing
-  → Host receives call, dispatches to Bootstrap Plugin A
-  → Plugin A sends BootstrapEvents via FfiBootstrapSender
-  → Host forwards events back to Plugin B via channel
+  → bootstrap_fn starts the provider on its own thread and returns an
+    FfiBootstrapStream immediately (FfiBootstrapReceiver for events +
+    FfiBootstrapResultReceiver for completion)
+  → Plugin B consumes both via BootstrapStreamConsumer / wrap_result_receiver
+    (plugin-sdk ffi/bootstrap_stream.rs); every link is bounded, so consumer
+    backpressure stalls the provider (issue #686), and cancellation is
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply lib/src/bootstrap/agents.md:31]: Done, took the suggested wording.

File: components/plugin-sdk/src/ffi/vtable_gen.rs [hunk 13]
```diff
@@ -2094,153 +2119,326 @@ pub fn build_reaction_vtable_from_boxed(
...
+        server_id: FfiStr,
+        source_id: FfiStr,
+    ) -> *mut FfiBootstrapStream {
+        ffi_guard(std::ptr::null_mut(), || {
+            use drasi_lib::bootstrap::{BootstrapContext, BootstrapRequest};
+
+            let request = BootstrapRequest {
+                query_id: unsafe { query_id.to_string() },
+                node_labels: (0..node_labels_count)
+                    .map(|i| unsafe { (*node_labels.add(i)).to_string() })
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/vtable_gen.rs:2384]: Done. Both arrays are null-guarded now (null with a non-zero count reads as empty), and there's a test covering exactly that call shape.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+use super::payload::consume_bootstrap_event;
+use super::vtables::{
+    FfiBootstrapEvent, FfiBootstrapReceiver, FfiBootstrapResult, FfiBootstrapResultReceiver,
+};
+
+/// Capacity of the bounded bridge between the producer's push callback and the
+/// async consumer. Mirrors the `sync_channel(256)` used by the CDC
+/// `ChangeReceiverProxy` and the source-plugin `BootstrapReceiverProxy`.
+pub const BOOTSTRAP_BRIDGE_CAPACITY: usize = 256;
+
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs:54]: Done. Both constants now live side by side in bootstrap_stream.rs with the aggregate in-flight bound documented on each.

File: components/plugin-sdk/src/ffi/vtable_gen.rs [hunk 15]
```diff
@@ -2094,153 +2119,326 @@ pub fn build_reaction_vtable_from_boxed(
...
     struct BootstrapProviderWrapper {
-        inner: Box<dyn BootstrapProvider>,
+        inner: Arc<dyn BootstrapProvider>,
     }
 
-    extern "C" fn bootstrap_fn(
+    /// Provider-side event capacity. Small by design: together with the
+    /// consumer bridge this caps in-flight events, so a stalled query stalls
+    /// the provider instead of buffering the snapshot.
+    const PROVIDER_EVENT_CAPACITY: usize = 100;
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/vtable_gen.rs]: Done. Exported as BOOTSTRAP_PROVIDER_CAPACITY next to the bridge capacity, vtable_gen uses it, and the tests import it instead of hardcoding 100.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+) {
+    let _ = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| drop_fn(state)));
+}
+
+/// Free a bootstrap event receiver without consuming its stream.
+///
+/// Error-path cleanup for a partially-initialized `FfiBootstrapStream`:
+/// releases the producer-side state via its `drop_fn` (which closes the
+/// provider's event channel, unblocking the provider) so nothing leaks.
+pub fn release_bootstrap_receiver(inner: FfiBootstrapReceiver) {
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs:184]: Done, added releasing_receivers_cancels_provider: drives the raw vtable, releases both receivers, and asserts the provider gets cancelled and never produced past its bounded channel.

File: components/plugin-sdk/src/ffi/bootstrap_stream.rs [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+        let tx = match wrapper.tx.lock() {
+            Ok(mut guard) => guard.take(),
+            Err(_) => None,
+        };
+        let Some(tx) = tx else {
+            return;
+        };
+        if result.is_null() {
+            let _ = tx.send(Err(anyhow::anyhow!(
+                "Bootstrap provider ended without a result"
```
[the remark this answers is not in the document]
Comment (amansinghoriginal) [reply components/plugin-sdk/src/ffi/bootstrap_stream.rs]: Done, added panicking_provider_yields_ended_without_result_error: the provider panics before producing a result, which drives the null-result path end to end through the real producer.

**Comment by amansinghoriginal** [comment]
> Host proxy `thread::spawn(...).join()` inside async still briefly pins a worker; OK if intentional for macOS TLS (document), else prefer `spawn_blocking`.

Intentional, and documented at the call site: constructing the consumers on the caller's thread would initialize plugin TLS there, and on macOS arm64 plugin TLS destructors can deadlock with the still-running plugin runtime at thread exit. Same rationale as the receiver proxies in SourceProxy::subscribe. The join is bounded to two constructor calls that each take a receiver out of a mutex, spawn a thread, and return, so the worker is held for microseconds, not for the bootstrap. Can switch to spawn_blocking if you still prefer.

Everything else is fixed on the current head or answered in its thread: blockers 1 and 2 rebutted in the threads (body locals vs captures, and both reclaims are once-guarded now so the concern is moot either way), should-fix 3, 4, 5 done, nits taken except the plateau rewrite (see thread, the suggested wait condition never fires when backpressure works). Ready for another look.

**Comment by amansinghoriginal** [comment]
@danielgerlag on blockers 1 and 2, here's a small self-contained test for the premise that a failed `spawn` fires the sentinel by dropping the closure. It shows the two halves separately: captures are dropped, body locals never run.

```rust
use std::sync::atomic::{AtomicBool, Ordering};

static CAPTURE_DROPPED: AtomicBool = AtomicBool::new(false);
static BODY_LOCAL_DROPPED: AtomicBool = AtomicBool::new(false);

struct Capture;
impl Drop for Capture {
    fn drop(&mut self) {
        CAPTURE_DROPPED.store(true, Ordering::SeqCst);
    }
}

/// Stand-in for SentinelOnDrop / CallOnce.
struct BodyLocal;
impl Drop for BodyLocal {
    fn drop(&mut self) {
        BODY_LOCAL_DROPPED.store(true, Ordering::SeqCst);
    }
}

#[test]
fn dropping_unrun_closure_drops_captures_not_body_locals() {
    let capture = Capture;
    let closure = move || {
        let _guard = BodyLocal; // constructed only if the closure runs
        let _keep = capture; // capture moved into the closure
    };
    drop(closure); // what happens when thread::Builder::spawn returns Err

    assert!(
        CAPTURE_DROPPED.load(Ordering::SeqCst),
        "captures ARE dropped (this is what a Drop counter measures)"
    );
    assert!(
        !BODY_LOCAL_DROPPED.load(Ordering::SeqCst),
        "body locals NEVER run their Drop: the sentinel cannot fire"
    );
}
```

Passes as written. A never-invoked FnOnce releases only its captured values; locals inside the body, which is where `SentinelOnDrop` and `CallOnce` live, are constructed only when the thread actually runs the body. So on spawn failure nothing fires from the drop, and the inline delivery in the Err arm is the single firing. I suspect the local verification put the counter on a capture, which increments in both worlds and so can't distinguish them.

That's why I kept the inline delivery (removing it would leave the consumer waiting forever on spawn failure) and added the once-guards on the reclaim paths, so even a hypothetical duplicate delivery is a no-op rather than a double free.

**Review by danielgerlag — APPROVED** [review]

**Review by agentofreality — APPROVED** [review]
